code/RealMeService.php

```diff
@@ -320,205 +465,291 @@ public function getBackURL()
     }
 
     /**
-     * @return string|null Either the directory where SimpleSAMLphp configuration is stored, or null if undefined
+     * @return string|null Either the directory where certificates are stored, or null if undefined
      */
-    public function getSimpleSamlConfigDir()
+    public function getCertDir()
     {
-        return (defined('REALME_CONFIG_DIR') ? rtrim(REALME_CONFIG_DIR, '/') : null);
+        return (defined('REALME_CERT_DIR') ? REALME_CERT_DIR : null);
     }
 
     /**
-     * @return string The path to SimpleSAMLphp's metadata. This will either be defined in config, or just '/metadata'
+     * Returns the appropriate AuthN Context, given the environment passed in. The AuthNContext may be different per
+     * environment, and should be one of the strings as defined in the static {@link self::$authn_contexts} at the top
+     * of this class.
+     *
+     * @param string $env The environment to return the AuthNContext for. Must be one of the RealMe environment names
+     * @return string|null Returns the AuthNContext for the given $env, or null if no context exists
      */
-    public function getSimpleSamlMetadataDir()
+    public function getAuthnContextForEnvironment($env)
     {
-        return sprintf('%s/metadata', $this->getSimpleSamlConfigDir());
+        return $this->getConfigurationVarByEnv('authn_contexts', $env);
     }
 
     /**
-     * @return string Either the value for baseurlpath in SimpleSAML's config, or a default value if it's been unset
+     * Returns the full path to the SAML signing certificate file, used by SimpleSAMLphp to sign all messages sent to
+     * RealMe.
+     *
+     * @return string|null Either the full path to the SAML signing certificate file, or null if it doesn't exist
      */
-    public function getSimpleSamlBaseUrlPath()
+    public function getSigningCertPath()
     {
-        if (strlen($this->config()->simplesaml_base_url_path) > 0) {
-            return $this->config()->simplesaml_base_url_path;
-        } else {
-            return 'simplesaml/';
-        }
+        return $this->getCertPath('SIGNING');
     }
 
-    /**
-     * @return string|null Either the directory where certificates are stored, or null if undefined
-     */
-    public function getCertDir()
+    public function getIdPCertPath()
     {
-        return (defined('REALME_CERT_DIR') ? REALME_CERT_DIR : null);
+        $cfg = $this->config();
+        $name = $this->getConfigurationVarByEnv('idp_x509_cert_filenames', $cfg->realme_env, $cfg->integration_type);
+        return Controller::join_links($this->getCertDir(), $name);
```

Comment:
@tractorcow says not to use `Controller` for anything filesystem related - use `realpath` and join them with vanilla string concatenation